File: command_parser.py
import re
from typing import Dict, Optional, List, Tuple
from desktop_controller import DesktopController
from automation_controller import KeyboardController, MouseController
from time_service import TimeService, ReminderManager
from weather_service import WeatherService

class CommandParser:
    """Parse natural language commands and execute desktop actions"""
    
    def __init__(self, weather_api_key: str = None):
        self.desktop = DesktopController()
        self.keyboard = KeyboardController()
        self.mouse = MouseController()
        self.time_service = TimeService()
        self.weather_service = WeatherService(api_key=weather_api_key)
        self.reminder_manager = ReminderManager()
        
        # Command patterns
        self.patterns = {
            # File operations
            'search_file': r'(?:search|find|look for|locate)\s+(?:file|files?)?\s*(?:named|called)?\s+(.+)',
            'open_file': r'open\s+(?:file\s+)?(.+)',
            'create_folder': r'create\s+(?:folder|directory)\s+(?:named|called)?\s+(.+)',
            
            # Application control
            'launch_app': r'(?:open|launch|start|run)\s+(.+?)(?:\s+app(?:lication)?)?$',
            'close_app': r'close\s+(.+?)(?:\s+app(?:lication)?)?$',
            
            # Window management
            'minimize_window': r'minimize\s+(.+?)(?:\s+window)?$',
            'maximize_window': r'maximize\s+(.+?)(?:\s+window)?$',
            'close_window': r'close\s+(?:window\s+)?(.+)',
            'focus_window': r'(?:focus|switch to|go to)\s+(.+?)(?:\s+window)?$',
            'list_windows': r'(?:list|show|what)\s+(?:windows|apps)\s+(?:are\s+)?(?:open|running)',
            
            # Keyboard actions
            'type_text': r'type\s+(.+)',
            'press_key': r'press\s+(.+?)(?:\s+key)?$',
            'execute_shortcut': r'(?:execute|do|press)\s+(.+?)\s+(?:shortcut|command)',
            
            # Common shortcuts
            'copy': r'copy(?:\s+(?:this|that|text))?$',
            'paste': r'paste(?:\s+(?:this|that|text))?$',
            'save': r'save(?:\s+(?:this|file))?$',
            'undo': r'undo(?:\s+(?:that|last))?$',
            
            # Mouse actions
            'click': r'click(?:\s+(?:here|there|mouse))?$',
            'double_click': r'double\s*click$',
            'right_click': r'right\s*click$',
            'scroll_up': r'scroll\s+up',
            'scroll_down': r'scroll\s+down',
            
            # System info
            'system_info': r'(?:system|computer)\s+(?:info|status|stats)',
            
            # Time queries
            'what_time': r'(?:what|current)\s+time',
            'what_date': r'(?:what|today\'s?)\s+(?:date|day)',
            'is_weekend': r'(?:is it|it\'s)\s+(?:the\s+)?weekend',
            
            # Weather queries
            'weather': r'(?:what\'s|what is|how\'s|how is|check|get)\s+(?:the\s+)?weather',
            'weather_in': r'weather\s+in\s+(.+)',
        }
    
    def parse_and_execute(self, command: str) -> Tuple[bool, str]:
        """
        Parse a natural language command and execute it
        
        Args:
            command: Natural language command string
            
        Returns:
            Tuple of (success, response_message)
        """
        command = command.strip().lower()
        
        # Try to match command patterns
        for action, pattern in self.patterns.items():
            match = re.search(pattern, command, re.IGNORECASE)
            if match:
                return self._execute_action(action, match)
        
        return False, "I didn't understand that command. Try something like 'open chrome' or 'search file report.pdf'"
    
    def _execute_action(self, action: str, match: re.Match) -> Tuple[bool, str]:
        """Execute a matched action"""
        
        try:
            # File operations
            if action == 'search_file':
                filename = match.group(1).strip()
                results = self.desktop.search_files(filename, limit=5)
                if results:
                    response = f"Found {len(results)} file(s):\n"
                    for r in results[:3]:
                        response += f"- {r['name']} at {r['path']}\n"
                    return True, response
                return False, f"No files found matching '{filename}'"
            
            elif action == 'open_file':
                filename = match.group(1).strip()
                # First try to search for the file
                results = self.desktop.search_files(filename, limit=1)
                if results:
                    success = self.desktop.open_file(results[0]['path'])
                    if success:
                        return True, f"Opened {results[0]['name']}!"
                return False, f"Couldn't find or open '{filename}'"
            
            elif action == 'create_folder':
                folder_name = match.group(1).strip()
                # Create in user's home directory by default
                import os
                path = os.path.join(os.path.expanduser('~'), 'Desktop', folder_name)
                success = self.desktop.create_folder(path)
                if success:
                    return True, f"Created folder '{folder_name}' on Desktop!"
                return False, f"Couldn't create folder '{folder_name}'"
            
            # Application control
            elif action == 'launch_app':
                app_name = match.group(1).strip()
                success = self.desktop.launch_application(app_name)
                if success:
                    return True, f"Launched {app_name}!"
                return False, f"Couldn't launch {app_name}"
            
            elif action == 'close_app':
                app_name = match.group(1).strip()
                success = self.desktop.close_application(app_name)
                if success:
                    return True, f"Closed {app_name}!"
                return False, f"Couldn't find or close {app_name}"
            
            # Window management
            elif action == 'minimize_window':
                window_title = match.group(1).strip()
                success = self.desktop.minimize_window(window_title)
                if success:
                    return True, f"Minimized {window_title}!"
                return False, f"Couldn't find window '{window_title}'"
            
            elif action == 'maximize_window':
                window_title = match.group(1).strip()
                success = self.desktop.maximize_window(window_title)
                if success:
                    return True, f"Maximized {window_title}!"
                return False, f"Couldn't find window '{window_title}'"
            
            elif action == 'close_window':
                window_title = match.group(1).strip()
                success = self.desktop.close_window(window_title)
                if success:
                    return True, f"Closed {window_title}!"
                return False, f"Couldn't find window '{window_title}'"
            
            elif action == 'focus_window':
                window_title = match.group(1).strip()
                success = self.desktop.focus_window(window_title)
                if success:
                    return True, f"Switched to {window_title}!"
                return False, f"Couldn't find window '{window_title}'"
            
            elif action == 'list_windows':
                windows = self.desktop.get_windows()
                if windows:
                    response = f"Open windows ({len(windows)}):\n"
                    for w in windows[:10]:  # Show first 10
                        response += f"- {w}\n"
                    return True, response
                return False, "No windows open"
            
            # Keyboard actions
            elif action == 'type_text':
                text = match.group(1).strip()
                success = self.keyboard.type_text(text)
                if success:
                    return True, f"Typed: {text}"
                return False, "Couldn't type text"
            
            elif action == 'press_key':
                key = match.group(1).strip()
                success = self.keyboard.press_key(key)
                if success:
                    return True, f"Pressed {key} key"
                return False, f"Couldn't press {key}"
            
            elif action == 'execute_shortcut':
                shortcut = match.group(1).strip()
                success = self.keyboard.execute_shortcut(shortcut)
                if success:
                    return True, f"Executed {shortcut} shortcut"
                return False, f"Unknown shortcut: {shortcut}"
            
            # Common shortcuts
            elif action == 'copy':
                success = self.keyboard.copy_text()
                return True, "Copied!" if success else False, "Couldn't copy"
            
            elif action == 'paste':
                success = self.keyboard.paste_text()
                return True, "Pasted!" if success else False, "Couldn't paste"
            
            elif action == 'save':
                success = self.keyboard.save()
                return True, "Saved!" if success else False, "Couldn't save"
            
            elif action == 'undo':
                success = self.keyboard.undo()
                return True, "Undone!" if success else False, "Couldn't undo"
            
            # Mouse actions
            elif action == 'click':
                success = self.mouse.click()
                return True, "Clicked!" if success else False, "Couldn't click"
            
            elif action == 'double_click':
                success = self.mouse.double_click()
                return True, "Double-clicked!" if success else False, "Couldn't double-click"
            
            elif action == 'right_click':
                success = self.mouse.right_click()
                return True, "Right-clicked!" if success else False, "Couldn't right-click"
            
            elif action == 'scroll_up':
                success = self.mouse.scroll_up()
                return True, "Scrolled up!" if success else False, "Couldn't scroll"
            
            elif action == 'scroll_down':
                success = self.mouse.scroll_down()
                return True, "Scrolled down!" if success else False, "Couldn't scroll"
            
            # System info
            elif action == 'system_info':
                info = self.desktop.get_system_info()
                response = "System Status:\n"
                response += f"CPU: {info.get('cpu_percent', 'N/A')}%\n"
                response += f"Memory: {info.get('memory_percent', 'N/A')}%\n"
                response += f"Disk: {info.get('disk_usage', 'N/A')}%\n"
                if info.get('battery'):
                    response += f"Battery: {info['battery']}%"
                return True, response
            
            # Time queries
            elif action == 'what_time':
                return True, self.time_service.get_smart_time_response("what time")
            
            elif action == 'what_date':
                return True, self.time_service.get_smart_time_response("what date")
            
            elif action == 'is_weekend':
                return True, self.time_service.get_smart_time_response("weekend")
            
            # Weather queries
            elif action == 'weather':
                return True, self.weather_service.get_smart_weather_response()
            
            elif action == 'weather_in':
                city = match.group(1).strip()
                return True, self.weather_service.get_smart_weather_response(city)
        
        except Exception as e:
            return False, f"Error executing command: {e}"
        
        return False, "Command not implemented yet"
    
    def is_desktop_command(self, text: str) -> bool:
        """Check if text contains a desktop control command"""
        text_lower = text.lower()
        
        command_keywords = [
            'open', 'close', 'launch', 'start', 'run',
            'search', 'find', 'look for',
            'minimize', 'maximize', 'focus', 'switch to',
            'type', 'press', 'click', 'scroll',
            'copy', 'paste', 'save', 'undo',
            'create folder', 'delete', 'move',
            'system info', 'system status',
            'what time', 'what date', 'weekend',
            'weather', 'temperature'
        ]
        
        return any(keyword in text_lower for keyword in command_keywords)

File: honey-rs/src/command_parser.rs
use crate::automation_controller::{automation_backend, KeyboardController, MouseController};
use crate::desktop_controller::DesktopController;
use crate::time_service;
use crate::weather_service::WeatherService;

pub struct CommandParser {
    desktop: DesktopController,
    keyboard: KeyboardController,
    mouse: MouseController,
    weather: WeatherService,
}

type Outcome = (bool, String);

impl CommandParser {
    pub fn new(weather: WeatherService) -> Self {
        Self {
            desktop: DesktopController::new(),
            keyboard: KeyboardController::new(),
            mouse: MouseController::new(),
            weather,
        }
    }

    pub fn is_desktop_command(text: &str) -> bool {
        let lower = text.to_lowercase();
        const KEYWORDS: &[&str] = &[
            "open", "close", "launch", "start", "run", "search", "find",
            "look for", "locate", "minimize", "maximize", "focus",
            "switch to", "go to", "type ", "press", "click", "scroll",
            "copy", "paste", "save", "undo", "redo", "cut", "select all",
            "create folder", "delete", "move", "system info", "system status",
            "what time", "what date", "today's date", "weekend", "weather",
            "temperature", "screenshot", "lock screen", "task manager",
            "windows",
        ];
        KEYWORDS.iter().any(|k| lower.contains(k))
    }

    pub fn parse_and_execute(&mut self, command: &str) -> Outcome {
        let cmd = command.trim().to_lowercase();

        let actions: &[(&str, &str)] = &[
            ("copy", r"^(?:please\s+)?copy(?:\s+(?:this|that|text))?$"),
            ("paste", r"^(?:please\s+)?paste(?:\s+(?:this|that|text))?$"),
            ("save", r"^(?:please\s+)?save(?:\s+(?:this|that|file))?$"),
            ("undo", r"^(?:please\s+)?undo(?:\s+(?:this|that|last))?$"),
            ("double_click", r"^double\s*click$"),
            ("right_click", r"^right\s*click$"),
            ("click", r"^(?:left\s+)?click(?:\s+(?:here|there|mouse))?$"),
            ("scroll_up", r"^scroll\s+up$"),
            ("scroll_down", r"^scroll\s+down$"),
            ("system_info", r"(?:system|computer)\s+(?:info|status|stats)"),
            ("what_time", r"(?:what(?:'s| is)?\s+)?(?:the\s+)?(?:current\s+)?time"),
            ("what_date", r"(?:what(?:'s| is)?\s+)?(?:today'?s?\s+)?date|(?:what\s+day\s+is\s+it)"),
            ("is_weekend", r"(?:is it|it's)\s+(?:the\s+)?weekend"),
            ("weather_in", r"weather\s+(?:in|for|at)\s+(.+)"),
            ("weather", r"(?:what(?:'s| is)|how(?:'s| is)|check|get)\s+(?:the\s+)?weather"),
            ("list_windows", r"(?:list|show)\s+(?:all\s+)?(?:the\s+)?(?:open\s+)?windows|what\s+windows\s+are\s+open"),
            ("minimize_window", r"^minimize\s+(?:the\s+)?window\s+(.+)$"),
            ("maximize_window", r"^maximize\s+(?:the\s+)?window\s+(.+)$|^maximize\s+(.+)$"),
            ("close_window", r"^close\s+(?:the\s+)?window\s+(.+)$|^close\s+window$"),
            ("focus_window", r"^(?:focus|switch to|go to)\s+(?:window\s+)?(.+)$"),
            ("execute_shortcut", r"^(?:execute|do|press)\s+(.+?)\s+(?:shortcut|command)$"),
            ("press_key", r"^press\s+(?:the\s+)?(.+?)(?:\s+key)?$"),
            ("create_folder", r"^create\s+(?:a\s+)?(?:folder|directory)\s+(?:named\s+|called\s+)?(.+)$"),
            ("search_file", r"^(?:search|find|look for|locate)\s+(?:for\s+)?(?:files?\s+)?(?:named\s+|called\s+)?(.+)$"),
            ("open_file_ext", r"^(?:open|show)\s+(?:the\s+)?(.+\.[a-z0-9]{1,5})$"),
            ("open_file", r"^(?:open|show)\s+(?:the\s+)?file\s+(.+)$"),
            ("close_app", r"^close\s+(.+?)(?:\s+app(?:lication)?)?$"),
            ("launch_app", r"^(?:open|launch|start|run)\s+(.+?)(?:\s+app(?:lication)?)?$"),
            ("type_text", r"^type\s+(.+)$"),
        ];

        for (action, pattern) in actions {
            let Ok(re) = regex::Regex::new(pattern) else { continue };
            if let Some(caps) = re.captures(&cmd) {
                return self.execute(action, &caps);
            }
        }

        (
            false,
            "I didn't understand that command. Try something like 'open firefox' or 'find report pdf'.".into(),
        )
    }

    fn execute(&self, action: &str, caps: &regex::Captures) -> Outcome {
        match action {
            "copy" => self.ok_or_err(
                self.keyboard.combo(&["ctrl", "c"]),
                "Copied!",
                "Couldn't copy",
            ),
            "paste" => self.ok_or_err(
                self.keyboard.combo(&["ctrl", "v"]),
                "Pasted!",
                "Couldn't paste",
            ),
            "save" => self.ok_or_err(
                self.keyboard.combo(&["ctrl", "s"]),
                "Saved!",
                "Couldn't save",
            ),
            "undo" => self.ok_or_err(
                self.keyboard.combo(&["ctrl", "z"]),
                "Undone!",
                "Couldn't undo",
            ),
            "click" => self.ok_or_err(self.mouse.click(), "Clicked!", "Couldn't click"),
            "double_click" => self.ok_or_err(self.mouse.double_click(), "Double-clicked!", "Couldn't double-click"),
            "right_click" => self.ok_or_err(self.mouse.right_click(), "Right-clicked!", "Couldn't right-click"),
            "scroll_up" => self.ok_or_err(self.mouse.scroll_up(), "Scrolled up!", "Couldn't scroll"),
            "scroll_down" => self.ok_or_err(self.mouse.scroll_down(), "Scrolled down!", "Couldn't scroll"),

            "system_info" => (true, self.desktop.system_info()),

            "what_time" => (true, time_service::smart_time_response("what time")),
            "what_date" => (true, time_service::smart_time_response("what day")),
            "is_weekend" => (true, time_service::smart_time_response("weekend")),

            "weather_in" => {
                let city = caps.get(1).map(|m| m.as_str().trim()).unwrap_or("");
                (true, self.weather.smart_response(Some(city)))
            }
            "weather" => (true, self.weather.smart_response(None)),

            "list_windows" => {
                let windows = self.desktop.list_windows();
                if windows.is_empty() {
                    (false, "No windows open".into())
                } else {
                    let mut resp = format!("Open windows ({}):\n", windows.len());
                    for w in windows.iter().take(10) {
                        resp.push_str(&format!("- {w}\n"));
                    }
                    (true, resp)
                }
            }

            "minimize_window" | "maximize_window" | "close_window" | "focus_window" => {
                // close_window with no capture means active window
                let title = caps
                    .get(1)
                    .map(|m| m.as_str().trim().to_string())
                    .unwrap_or_default();
                let result = match action {
                    "minimize_window" => self
                        .desktop
                        .minimize_window(&title)
                        .map(|_| format!("Minimized {}!", title)),
                    "maximize_window" => self
                        .desktop
                        .maximize_window(&title)
                        .map(|_| format!("Maximized {}!", title)),
                    "close_window" => {
                        if title.is_empty() {
                            Err(anyhow::anyhow!("no window name given"))
                        } else {
                            self.desktop
                                .close_window(&title)
                                .map(|_| format!("Closed {}!", title))
                        }
                    }
                    _ => self
                        .desktop
                        .focus_window(&title)
                        .map(|_| format!("Switched to {}!", title)),
                };
                match result {
                    Ok(msg) => (true, msg),
                    Err(e) => (false, format!("{e}")),
                }
            }

            "execute_shortcut" => {
                let name = caps.get(1).map(|m| m.as_str()).unwrap_or("").to_string();
                match self.keyboard.execute_shortcut(&name) {
                    Ok(()) => (true, format!("Executed {name} shortcut")),
                    Err(e) => (false, format!("{e}")),
                }
            }

            "press_key" => {
                let key = caps.get(1).map(|m| m.as_str()).unwrap_or("").trim().to_string();
                match self.keyboard.press_key(&key) {
                    Ok(()) => (true, format!("Pressed {key} key")),
                    Err(e) => (false, format!("Couldn't press {key}: {e}")),
                }
            }

            "type_text" => {
                let text = caps.get(1).map(|m| m.as_str()).unwrap_or("").to_string();
                match self.keyboard.type_text(&text) {
                    Ok(()) => (true, format!("Typed: {text}")),
                    Err(e) => (false, format!("Couldn't type: {e}")),
                }
            }

            "create_folder" => {
                let name = caps.get(1).map(|m| m.as_str()).unwrap_or("").to_string();
                match self.desktop.create_folder(&name) {
                    Ok(p) => (true, format!("Created folder '{}' at {}", name, p.display())),
                    Err(e) => (false, format!("Couldn't create folder '{name}': {e}")),
                }
            }

            "search_file" => {
                let query = caps.get(1).map(|m| m.as_str().trim()).unwrap_or("");
                let results = self.desktop.search_files(query, 5);
                if results.is_empty() {
                    (false, format!("No files found matching '{query}'"))
                } else {
                    let mut resp = format!("Found {} file(s):\n", results.len());
                    for r in results.iter().take(3) {
                        resp.push_str(&format!("- {} at {}\n", r.name, r.path));
                    }
                    (true, resp)
                }
            }

            "open_file_ext" | "open_file" => {
                let filename = caps.get(1).map(|m| m.as_str()).unwrap_or("");
                let results = self.desktop.search_files(filename, 1);
                if results.is_empty() {
                    (false, format!("Couldn't find a file matching '{filename}'"))
                } else {
                    match self.desktop.open_file(&results[0].path) {
                        Ok(()) => (true, format!("Opened {}!", results[0].name)),
                        Err(e) => (false, format!("Couldn't open {}: {e}", results[0].name)),
                    }
                }
            }

            "close_app" => {
                let app = caps.get(1).map(|m| m.as_str().trim()).unwrap_or("").to_string();
                if app == "window" {
                    return (false, "Which window should I close?".into());
                }
                if self.desktop.close_application(&app) {
                    (true, format!("Closed {app}!"))
                } else {
                    (false, format!("Couldn't find or close {app}"))
                }
            }

            "launch_app" => {
                let app = caps.get(1).map(|m| m.as_str().trim()).unwrap_or("").to_string();
                match self.desktop.launch_application(&app) {
                    Ok(msg) => (true, msg),
                    Err(e) => (false, format!("{e}")),
                }
            }

            _ => (
                false,
                format!("Command not implemented yet (backend: {})", automation_backend()),
            ),
        }
    }

    fn ok_or_err(&self, result: anyhow::Result<()>, ok_msg: &str, err_prefix: &str) -> Outcome {
        match result {
            Ok(()) => (true, ok_msg.to_string()),
            Err(e) => (false, format!("{err_prefix}: {e}")),
        }
    }
}
